saw-core/src/Verifier/SAW/Term/Functor.hs

```diff
@@ -132,6 +133,50 @@ maxSort :: [Sort] -> Sort
 maxSort [] = propSort
 maxSort ss = maximum ss
 
+-- | This type represents a set of advisory flags for 'Sort's that are mostly
+-- ignored, but are used in the Coq export process to indicate where various
+-- typeclass instances are necessary in function definitions. In the concrete
+-- syntax "isort", "qsort", etc. is used to indicate cases where these flags
+-- are set. Note in particular that this flag does not affect typechecking,
+-- so missing or overeager "isort"/"qsort" annotations will only be detected
+-- via the Coq export.
+--
+-- * If 'flagInhabited' is 'True', an implicit @Inhabited@ typeclass argument
+--   will be added during Coq translation. In the concrete syntax, an "i" is
+--   prepended to the sort (e.g. "isort").
+-- * If 'flagQuantType' is 'True', an implicit @QuantType@ typeclass argument
+--   will be added during Coq translation. In the concrete syntax, an "q" is
+--   prepended to the sort (e.g. "qsort", "qisort").
+data SortFlags = SortFlags { flagInhabited :: Bool
+                           , flagQuantType :: Bool }
+    deriving (Eq, Ord, Generic, TH.Lift)
+
+instance Hashable SortFlags -- automatically derived
+
+instance Show SortFlags where
+  showsPrec _ (SortFlags i q) =
+    showString $ (if q then "q" else "") ++
+                 (if i then "i" else "")
+
+-- | The 'SortFlags' object with no flags set
+noFlags :: SortFlags
+noFlags = SortFlags False False
+
+-- | Combine two 'SortFlags' by setting each flag in the result if and only if
+-- it flag is set in both inputs
+cbnFlags :: SortFlags -> SortFlags -> SortFlags
```

Comment:
"cbnFlags" is an unfortunate name, because it sounds like we are talking about call-by-name. Maybe just use "combineFlags"? Or, better, maybe use a `Monoid` instance?

Reply:
First of all I should've addressed your `Monoid` point: `noFlags` and `cbnFlags` don't form a monoid! To be a monoid the `mempty` element needs to be an identity, and for `cbnFlags`, it's `SortFlags True True` which is the identity, not `noFlags = SortFlags False False`. (Sorry for not stating that earlier.)

My idea was that to avoid that potential confusion (since if we define `(<>)` it would seem logical that `noFlags` should be its `mempty`) we should just be explicit about what operation is being applied there, thus `liftA2 (&&)` instead of an opaque `cbnFlags` or `(<>)`.

That being said, instead of making `SortFlags` a functor I could have just defined a `sortFlagsBinOp` function or something which did the same thing as `liftA2`. If you'd prefer that, I will add back that function, add back `sortFlagsToList`, and get rid of `SortFlagsF`. But, I personally prefer it this way since I just generally prefer the prelude `liftA2`, `toList`, etc. to custom `sortFlagsBinOp`, `sortFlagsToList` functions (plus having the fully general `fmap`, `liftA2` meant I could simplify some of the definitions in that file).

Of course if we do keep it, I will make it clearer in the Haddock that `Bool` is used everywhere else in the codebase, and that it's just a functor for convenience.

Let me know what you prefer. Sorry for not communicating my thinking here more clearly!

Reply:
Ah, ok. If it's not really an identity, then you shouldn't make it a monoid. :)

I see what you mean about it being nice to combine flags pointwise, and how making `SortFlags` an applicative functor lets you do that, but I still think it's confusing to make that a type variable, because that's not really part of the intent of this type. You should think about it as an API, or an abstraction: when a client of the `SortFlags` type wants to use it, what operations and interface should they see? I would argue a `SortFlags` represents a collection of Boolean flags, that can be combined (what is the semantics of combining them? That should be in the Haddock) and each of which can be initialized. Your API to `SortFlags` should reflect that usage pattern. Does that make sense?

Reply:
No, that makes sense – I think I just got so preoccupied with whether or not I could make it a functor I didn't stop to think if I should. 🙂 Changed `SortFlagsF` back to `SortFlags`, with now a `sortFlagsLift2` operation that I use for that one spot (i.e. as `sortFlagsLift2 (&&)`).